cmd/controller/controller.go

```diff
@@ -155,6 +171,83 @@ func (c *Controller) processNextItem() bool {
 	return true
 }
 
+func fetchUrl(reqURL, authHeader string) (*http.Response, error) {
+	req, err := http.NewRequest("GET", reqURL, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(authHeader) > 0 {
+		req.Header.Set("Authorization", authHeader)
+	}
+	return netClient.Do(req)
+}
+
+func fetchRepoIndex(repoURL string, authHeader string) (*repo.IndexFile, error) {
+	parsedURL, err := url.ParseRequestURI(strings.TrimSpace(repoURL))
+	if err != nil {
+		return nil, err
+	}
+	parsedURL.Path = strings.TrimSuffix(parsedURL.Path, "/") + "/index.yaml"
+
+	res, err := fetchUrl(parsedURL.String(), authHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return nil, errors.New("repo index request failed")
+	}
+
+	defer res.Body.Close()
```

Comment:
This defer should go before the `StatusCode` test above - otherwise you aren't closing the body for non-200 responses.

Reply:
It appears `res.Body.Close()` should be called whenever `res` is not `nil`. updated!

Reply:
for reference the sidenote on https://blog.learngoprogramming.com/5-gotchas-of-defer-in-go-golang-part-iii-36a1ab3d6ef1?token=ykNJIL3gxrqkXrMJ regarding this subject says: 

> _Here, you also need to check whether resp is nil, this is a caveat for http.Get. Usually, when there is an error, the response will be nil and an error will be returned. But, when you get a redirection error, response will not be nil but there’ll be an error. With the above code, you’re ensuring that you’re closing the response body. You also need to discard the data received if you’re not going to use it._

Reply:
The next paragraph on that same blog post says:
> UPDATE: This problem looks like doesn’t exist with http. I need to find a better example for that. So, it may still be valid for some code, but not for http. Check out this discussion: https://medium.com/@mafredri/great-article-thanks-c0e88d4df19e. Thanks Mathias Fredriksson and Patrycja Szabłowska for validating Mathias’s point. Check out it here: https://medium.com/@szablowska.patrycja/thanks-for-the-article-bdcca5eda295.

The updated conclusion is that unconditionally calling `defer res.Body.Close()` in the non-error case is correct in all cases, and more idiomatic.  (ie: `res.Body` doesn't need to be closed if err != nil, and `res.Body` is always non-nil if err is nil.)